### ISSUE TITLE
Batch rendering using gaussian rasterizer

### ISSUE THREAD

**Issue by AlleUndKalle** [graphdeco-inria/gaussian-splatting#92]
Hi,

Thanks for open-sourcing this awesome work! I noticed that the differentiable rasterizer only supports single image rendering at a time. I wonder would it be possible to perform batch rendering assuming we have same number of gaussians with different properties for each frame? If not, what would be the required changes to be made? I am willing to work on it if you can give some pointers on how to do that.

**Comment by jaeseok4104** [comment]
i think, we can optimize using several image. i agree your opinion

**Comment by Linkersem** [comment]
Yes, i think using several image will be very interesting

**Comment by yzslab** [comment]
I think you can pop muitiple images, render them and then average their loss in each iterations: https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/train.py#L78
Or try to use gradient accumulation: https://github.com/graphdeco-inria/gaussian-splatting/blob/f7a116fb1397d9842239127d39dc212f93171f70/train.py#L124

**Comment by grgkopanas** [comment]
Given internal tests that we did the last answer should work

**Comment by Snosixtyboo** [comment]
As @yzslab pointed out, you can accumulate gradients over multiple forwards/backwards and then step. If you want true batching that uses the GPU more, I would recommend going with cuda streams. Off the top of my head, you would need to pass in multiple views (cameras) that you want to do simultaneously, and then on the CPP/CUDA side:
- create multiple streams (e.g. 4)
- make duplicates in the forward of the auxiliary structures (geomBuffer, binningBuffer, imgBuffer) for each stream
- make duplicates in the backward of the gradient tensors (dL_d...) for each stream
- run forward/backward using one stream per view

I'm sure I'm forgetting something and it's not as trivial as it sounds here :)

**Comment by BitCalSaul** [comment]
@Snosixtyboo Hi, thanks for your insights. I’m just curious — does 3DGS rendering (both forward and backward) natively support batching like neural networks do? If not, then increasing the batch size likely wouldn’t offer any speedup, right?
